File: Module 4 - Data Structure/Array/Array.py
""" Python's not have built-in array so we use list which is dynamic and can hold different object types, 
so it is not exactly the same as a C++ fixed array.
Python lists can grow dynamically.

keyword
* len() - returns the number of items in a list
* append() - adds an item to the end of the list
* insert() - adds an item at the specified index
* delete() - removes an item at the specified index
* pop() - removes and returns the item at the specified index
"""

arr = [10, 20, 30, 40, 50]  # list of size 5

'''
Index:   0   1   2   3   4
         ↓   ↓   ↓   ↓   ↓
Values:  10  20  30  40  50
'''
# Operation on List
# Traversal
print("Traversal of List elements: ")
for x in arr:
    print(x)

print()

# Accessing an element
print("Accessing an element at index 2: ")
print(arr[2])  # Output: 30
print()

# Updating an element
print("Updating an element at index 3: ")
arr[3] = 100
print(arr[3])  # Output: 100
print()

# searching an element
print("Searching for an element (30): ")
target = 30
for i in range(len(arr)):
    if arr[i] == target:
        print("Found at index", i)
        break

print()

# insert an element
print("Inserting an elemen at index 2: ")
arr.insert(2, 25)
print(arr)
print()

# delete an element
print("Deleting an element at index 2: ")
del arr[2]
print(arr)

# add an element at the end
print("Adding an element at the end: ")
arr.append(60)
print(arr)
print()

# removing an element at a specific index
print("Removing an element at index 1: ")
arr.pop(1)
print(arr)
print()

# list with different data types
mixed_list = [1, "Hello", 3.14, True]
print("List with different data types: ", mixed_list)

File: Module 4 - Data Structure/Array/Array.cpp
/* Traditional C++ arrays have a fixed size.
<datatype> <array_name>[<size>] = {<value1>, <value2>, <value3>, ...};

operations on arrays:
1. size of array - sizeof(array_name)/sizeof(array_name[0])
* sizeof(array_name) gives the total size of the array in bytes.
* sizeof(array_name[0]) gives the size of a single element in the array in bytes

For using specific Operations on arrays,
use array library,
* #include<array>
*/

# include<iostream>
    using namespace std;

    int main(){
        
int arr[5] = {10, 20, 30, 40, 50};  // array of size 5 

/*
Index:   0   1   2   3   4
         ↓   ↓   ↓   ↓   ↓
Values:  10  20  30  40  50
*/

// Operation on Array
// size of array
cout << "Size of the array: " << endl;
cout << sizeof(arr)/sizeof(arr[0]) << endl;  // Output: 5

// Traversal
cout << "\nTraversal of Array elements: " << endl;
for(int i = 0; i < 5; i++){
    cout << arr[i] << endl;
}

// Accessing an element
cout << "\nAccessing an element at index 2: " << endl;
cout << arr[2] << endl;  // Output: 30

// Updating an element
cout << "\nUpdating an element at index 3: " << endl;
arr[3] = 100;  // Update the value at index 3
cout << arr[3] << endl;  // Output: 100

// searching an element
cout << "\nSearching for an element (20): " << endl;
int searchValue = 20;
for(int i = 0; i < 5; i++){
    if(arr[i] == searchValue){
        cout << "Element found at index: " << i << endl;
        break;
    }
}

// inserting an element (Note: In traditional arrays, you cannot change the size, so we can only update an existing index)
cout << "\nInserting an element (25) at index 2: " << endl;
int Arr[10] = {10, 20, 30, 40};
int n = 4;
int pos = 2;
int value = 25;

for(int i = n; i > pos; i--)
{
    Arr[i] = Arr[i - 1];
}

Arr[pos] = value;
n++;

cout << "Array after insertion: ";
for(int i = 0; i < n; i++){
    cout << Arr[i] << " ";
}

cout << endl;

// deleting an element (Note: In traditional arrays, you cannot change the size, so we can only shift elements to "remove" an element)n = 5;
cout << "\nDeleting an element at index 2: " << endl;
int Pos = 2;
for(int i = Pos; i < n - 1; i++)
{
    Arr[i] = Arr[i + 1];
}

n--;

cout << "Array after deletion: ";
for(int i = 0; i < n; i++){
    cout << Arr[i] << " ";
}

return  0;
}